Kerberos Authentication broken in OpenJDK 1.8 because of PR #373

The current master contains code from PR #373 that breaks Kerberos authentication in OpenJDK 1.8.

Initialization of `com.hierynomus.smbj.auth.ExtendedGSSContext` fails because `com.sun.security.jgss.InquireType` cannot be cast to `com.sun.security.jgss.ExtendedGSSContext`.

```
java.lang.ExceptionInInitializerError
	at com.hierynomus.smbj.auth.SpnegoAuthenticator.authenticateSession(SpnegoAuthenticator.java:96)
	[...]
Caused by: java.lang.ClassCastException: Cannot cast com.sun.security.jgss.InquireType to com.sun.security.jgss.ExtendedGSSContext
	at java.lang.Class.cast(Class.java:3369)
	at java.lang.invoke.MethodHandles.insertArguments(MethodHandles.java:2378)
	at com.hierynomus.smbj.auth.ExtendedGSSContext.getKrb5GetSessionKeyFunction(ExtendedGSSContext.java:52)
	at com.hierynomus.smbj.auth.ExtendedGSSContext.<clinit>(ExtendedGSSContext.java:29)
     [...]
```

It seems to me that this could be fixed by changing the `pos` argument in [`ExtendedGssContext. getKrb5GetSessionKeyFunction()`](https://github.com/hierynomus/smbj/commit/bd1e608c986339793481a48b45ecb21b91a3d019#diff-e3c786664caa6ff023501f16ea4ee075R52) from

`return MethodHandles.insertArguments(handle, 0, getSessionKeyConst).asType(MethodType.methodType(Key.class, GSSContext.class));`

to
`return MethodHandles.insertArguments(handle, 1, getSessionKeyConst).asType(MethodType.methodType(Key.class, GSSContext.class));`

because apparently, in OpenJDK, the class of a method is listed as a parameter type when listing the paramter types using the MethodHandles.insertArguments method:

<img width="1283" alt="MethodHandle insertArguments openjdk" src="https://user-images.githubusercontent.com/3786617/55864811-82805980-5b7d-11e9-9d15-e1e369441f7b.png">

I am currently working around this by catching exceptions in `SpnegoAuthenticator`

```
Key key;
try {
   key = ExtendedGSSContext.krb5GetSessionKey(gssContext);
} catch (ExceptionInInitializerError | NoClassDefFoundError ex) {
   com.sun.security.jgss.ExtendedGSSContext e = (com.sun.security.jgss.ExtendedGSSContext) gssContext;
   key = (Key) e.inquireSecContext(InquireType.KRB5_GET_SESSION_KEY);
}
```

Now, this is hardly a good solution. If anybody can point me in the right direction, I'm happy to provide a PR.

--- a/src/main/java/com/hierynomus/smbj/auth/ExtendedGSSContext.java
+++ b/src/main/java/com/hierynomus/smbj/auth/ExtendedGSSContext.java
@@ -22,10 +22,10 @@ import java.lang.reflect.Method;
 import java.security.Key;
 
 class ExtendedGSSContext {
-    private static final Method krb5GetSessionKey = getKrb5GetSessionKeyFunction();
-    private static Object getSessionKeyConst;
+    private static final Method inquireSecContext = getInquireSecContextMethod();
+    private static Object krb5GetSessionKeyConst;
 
-    private static Method getKrb5GetSessionKeyFunction() {
+    private static Method getInquireSecContextMethod() {
         Class<?> extendedContextClass;
         Class<?> inquireTypeClass;
         try {
@@ -42,7 +42,7 @@ class ExtendedGSSContext {
                 throw exception;
             }
         }
-        getSessionKeyConst = Enum.valueOf(inquireTypeClass.asSubclass(Enum.class), "KRB5_GET_SESSION_KEY");
+        krb5GetSessionKeyConst = Enum.valueOf(inquireTypeClass.asSubclass(Enum.class), "KRB5_GET_SESSION_KEY");
         try {
             return extendedContextClass.getDeclaredMethod("inquireSecContext", inquireTypeClass);
         } catch (NoSuchMethodException e) {
@@ -52,7 +52,7 @@ class ExtendedGSSContext {
 
     public static Key krb5GetSessionKey(GSSContext gssContext) throws TransportException {
         try {
-            return (Key) krb5GetSessionKey.invoke(gssContext, getSessionKeyConst);
+            return (Key) inquireSecContext.invoke(gssContext, krb5GetSessionKeyConst);
         } catch (Throwable e) {
             throw new TransportException(e);
         }

--- a/src/main/java/com/hierynomus/smbj/auth/ExtendedGSSContext.java
+++ b/src/main/java/com/hierynomus/smbj/auth/ExtendedGSSContext.java
@@ -15,20 +15,17 @@
  */
 package com.hierynomus.smbj.auth;
 
+import com.hierynomus.protocol.transport.TransportException;
 import org.ietf.jgss.GSSContext;
 
-import java.lang.invoke.MethodHandle;
-import java.lang.invoke.MethodHandles;
-import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
 import java.security.Key;
 
-
-import com.hierynomus.protocol.transport.TransportException;
-
 class ExtendedGSSContext {
-    private static final MethodHandle krb5GetSessionKey = getKrb5GetSessionKeyFunction();
+    private static final Method krb5GetSessionKey = getKrb5GetSessionKeyFunction();
+    private static Object getSessionKeyConst;
 
-    private static MethodHandle getKrb5GetSessionKeyFunction() {
+    private static Method getKrb5GetSessionKeyFunction() {
         Class<?> extendedContextClass;
         Class<?> inquireTypeClass;
         try {
@@ -45,23 +42,22 @@ class ExtendedGSSContext {
                 throw exception;
             }
         }
-        @SuppressWarnings("unchecked")
-        Object getSessionKeyConst = Enum.valueOf(inquireTypeClass.asSubclass(Enum.class), "KRB5_GET_SESSION_KEY");
+        getSessionKeyConst = Enum.valueOf(inquireTypeClass.asSubclass(Enum.class), "KRB5_GET_SESSION_KEY");
         try {
-            MethodHandle handle = MethodHandles.lookup().findVirtual(extendedContextClass, "inquireSecContext", MethodType.methodType(Object.class, inquireTypeClass));
-            return MethodHandles.insertArguments(handle, 0, getSessionKeyConst).asType(MethodType.methodType(Key.class, GSSContext.class));
-        } catch (NoSuchMethodException | IllegalAccessException e) {
+            return extendedContextClass.getDeclaredMethod("inquireSecContext", inquireTypeClass);
+        } catch (NoSuchMethodException e) {
             throw new IllegalStateException(e);
         }
     }
 
     public static Key krb5GetSessionKey(GSSContext gssContext) throws TransportException {
         try {
-            return (Key) krb5GetSessionKey.invokeExact(gssContext);
+            return (Key) krb5GetSessionKey.invoke(gssContext, getSessionKeyConst);
         } catch (Throwable e) {
             throw new TransportException(e);
         }
     }
 
-    private ExtendedGSSContext() {}
+    private ExtendedGSSContext() {
+    }
 }

